retry: fix retry.WithMaxAttempt to deal with opt.Closer properly

otan opened the pull request:
In `beac4a53e0e2e2236eb5957f67abc1bf476ad1b6`, we introduced
stopper.ShouldQuiesce() to the retry.Closer so that server shutdowns
also shut down in-process retries to the temp schema cleaner.

However, when stopper.ShouldQuiesce() is called, the error that gets
wrapped in `errors.Wrap` is nil (as ctx.Err() is nil), and as such we
return with no error set. This causes potentially bugs afterwards as
users of the functions expected errors when this happens and not to
continue silently.

This PR bridges that gap by always wrapping an error around cases where
WithMaxAttempt is aborted by a context attempt.

Resolves https://github.com/cockroachdb/cockroach/issues/47057.

Release note: None.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/47063)
<!-- Reviewable:end -->

otan committed:
retry: fix retry.WithMaxAttempt to deal with opt.Closer properly

In `beac4a53e0e2e2236eb5957f67abc1bf476ad1b6`, we introduced
stopper.ShouldQuiesce() to the retry.Closer so that server shutdowns
also shut down in-process retries to the temp schema cleaner.

However, when stopper.ShouldQuiesce() is called, the error that gets
wrapped in `errors.Wrap` is nil (as ctx.Err() is nil), and as such we
return with no error set. This causes potentially bugs afterwards as
users of the functions expected errors when this happens and not to
continue silently.

This PR bridges that gap by always wrapping an error around cases where
WithMaxAttempt is aborted by a context attempt.

Resolves https://github.com/cockroachdb/cockroach/issues/47057.

Release note: None.

knz commented:
Woah good catch!

@tbg this may be of interest. 
@otan @andreimatei I think this might explain #47011 - what do you think?

knz commented on pkg/util/retry/retry_test.go:
```diff
@@ -140,26 +142,148 @@ func TestRetryNextCh(t *testing.T) {
+			ctx:  context.Background(),
+			opts: Options{
+				InitialBackoff: time.Microsecond * 10,
+				MaxBackoff:     time.Second,
```
I would recommend reducing the backoff time - you don't need a full second of testing to know the test is successful.

otan replied:
fixed!

tbg commented:
Yikes, good catch @otan.

tbg commented:
Does this affect release-20.1?

beac4a53e0e2e2236eb5957f67abc1bf476ad1b6 is about the temporary object cleaner which I think caused the delay, so I'm hopeful that fixes that, too.

otan commented:
Yeah I will backport this along with the original fix.

On Mon, 6 Apr 2020, 3:38 am Tobias Grieger, <notifications@github.com>
wrote:

> Does this affect release-20.1?
>
> beac4a5
> <https://github.com/cockroachdb/cockroach/commit/beac4a53e0e2e2236eb5957f67abc1bf476ad1b6>
> is about the temporary object cleaner which I think caused the delay, so
> I'm hopeful that fixes that, too.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/47063#issuecomment-609715799>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA32FQ7ZYPSVIQOCJOETPMLRLGWLZANCNFSM4MB7CWAQ>
> .
>

otan commented:
> Does this affect release-20.1?

The original fix does, so when I backport it I will backport this PR along with it.

knz reviewed (approved):
thanks!

otan commented:
bors r=knz

RaduBerinde commented:
@knz I encountered this pattern before - `if err1 == nil { return err2 } else { return Wrap() }`. I think it would be useful to have an `errors` helper for this, perhaps something that takes a variable list of errors (some of which can be `nil`) and returns a single error. Or perhaps make Wrap specially handle a `nil` input.

knz commented:
You may be thinking about `errors.CombineErrors(err1, err2)`

RaduBerinde commented:
Oh, TIL, thanks!

andreimatei commented:
Yeah, I'm a big fan of `errors.CombineErrors()`. Real fine function.
